syntax = "proto3";

package transaction;

message NewOrder {
  repeated NewOrderMsg msgs = 1;
  repeated StdSignature signatures = 2;
  string memo = 3;
  int64 source = 4;
  bytes data = 5;
}
message CancelOrder{
  repeated CancelOrderMsg msgs=1;
  repeated StdSignature signatures = 2;
  string memo = 3;
  int64 source = 4;
  bytes data = 5;

}
message Send{
  repeated SendMsg msgs=1;
  repeated StdSignature signatures = 2;
  string memo = 3;
  int64 source = 4;
  bytes data = 5;
}
message TokenFreeze{
  repeated FreezeMsg msgs=1;
  repeated StdSignature signatures = 2;
  string memo = 3;
  int64 source = 4;
  bytes data = 5;
}
message TokenUnfreeze{
  repeated UnfreezeMsg msgs=1;
  repeated StdSignature signatures = 2;
  string memo = 3;
  int64 source = 4;
  bytes data = 5;
}
message Vote{
  repeated VoteMsg msgs=1;
  repeated StdSignature signatures = 2;
  string memo = 3;
  int64 source = 4;
  bytes data = 5;
}
message NewOrderMsg {
  bytes sender = 1; // order originating address
  string id = 2; // order id, this is much, please check the details below Order ID section.
  string symbol = 3; // symbol for trading pair in full name of the tokens
  int64 ordertype = 4; // only accept 2 for now, meaning limit order
  int64  side = 5; // 1 for buy and 2 fory sell
  int64 price = 6; // price of the order, which is the real price multiplied by 1e8 (10^8) and rounded to integer
  int64 quantity = 7; // quantity of the order, which is the real price multiplied by 1e8 (10^8) and rounded to integer
  int64 timeinforce = 8; // 1 for Good Till Expire(GTE) order and 3 for Immediate Or Cancel (IOC)

}
message CancelOrderMsg{
  bytes sender = 1; // order originating address'
  string symbol = 2; // symbol for trading pair in full name of the tokens
  string refid = 4; // order id of the one to cancel
}
message SendMsg{
  message Token {
    string denom = 1;
    int64 amount = 2;
  }
  message Input {
    bytes address = 1;
    repeated Token coins = 2;
  }
  message Output {
    bytes address = 1;
    repeated Token coins = 2;
  }
  repeated Input inputs = 1;
  repeated Output outputs = 2;
}
message FreezeMsg{
  bytes from = 1; // owner address
  string symbol = 2; // token symbol, in full name with "-" suffix
  int64 amount = 3; // amount of token to freeze
}
message UnfreezeMsg{
  bytes from = 1; // owner address
  string symbol = 2; // token symbol, in full name with "-" suffix
  int64 amount = 3; // amount of token to unfreeze
}
message VoteMsg{
  int64 proposal_id = 1;
  bytes voters = 2;
  VoteOption option = 3;
}

enum VoteOption {
  Yes = 0;
  Abstain = 1;
  No = 2;
  NoWithVeto = 3;
}

message StdSignature{
  message PubKey{
  }
  PubKey pub_key = 1;
  bytes signature = 2;
  int64 account_number = 3;
  int64 sequence = 4;
}
